### PR TITLE
Refactored klog.sh and added support for multi-container pods

### PR DESCRIPTION
**Pull request by bikramnehra**
This PR contains the following changes [(Trello Task)](https://trello.com/c/4KTbZVtU/901-2-klogsh-needs-to-be-updated-to-supported-multiple-containers-per-pod):

1. Updating `kubectl` commands 
2. Adding support for fetching kube logs from multi-container pods
2. Code Refactoring

### CODE REVIEW

File: container-host-files/opt/scf/bin/klog.sh [hunk 1]
```diff
@@ -5,192 +5,82 @@ set -e
...
 
-if [ "${FORCE}" == "1" ] ; then
-    rm -f "${DONE}" 2> /dev/null
+if [ "${FORCE}" == "1" ]; then
+	rm -f "${DONE}" 2>/dev/null
 fi
 
 function get_phase() {
-  kubectl get pod "${POD}" --namespace "${NS}" --output=jsonpath='{.status.phase}'
+	kubectl get pod "${POD}" -n "${NS}" -o jsonpath='{.status.phase}'
```
Comment (mook-as) [inline container-host-files/opt/scf/bin/klog.sh]: Please don't use short options in scripts; it's (typically) harder to see at a glance what they're doing if it's a tool that we haven't used manually in a while. (This probably doesn't apply to `kubectl`, but the standard style is long argument names in scripts)

File: container-host-files/opt/scf/bin/klog.sh [hunk 2]
```diff
@@ -5,192 +5,82 @@ set -e
 KLOG=${HOME}/klog
 
 if [ "$1" == "-h" ]; then
-    cat <<EOF
+	cat <<EOF
```
Comment (mook-as) [inline container-host-files/opt/scf/bin/klog.sh]: Not sure changing to tab indents (and other random stylistic changes) in an unrelated commit is a good idea.

File: container-host-files/opt/scf/bin/klog.sh [hunk 3]
```diff
@@ -5,192 +5,82 @@ set -e
...
-        # Get the pod logs - previous may not be there if it was successful on the first run.
-        # Unfortunately we can't get anything past the previous one
-        kubectl logs "${POD}" --namespace "${NS}" > "${DIR}/kube.log"
-        kubectl logs "${POD}" --namespace "${NS}" --previous > "${DIR}/kube-previous.log" 2> /dev/null || true
-        kubectl describe pods "${POD}" --namespace "${NS}" > "${DIR}/describe-pod.txt"
-    done
+		# Get the CF logs inside the pod if there are any
+		if [ "$(get_phase)" != 'Succeeded' ]; then
+			if [ check_for_log_dir ]; then
+				kubectl cp -n "${NS}" "${POD}":/var/vcap/sys/log/ "${POD_DIR}/" 2>/dev/null
```
Comment (mook-as) [inline container-host-files/opt/scf/bin/klog.sh]: `set -e` (line 3 of the script) should still be active; so the old `|| true` was to temporarily disable that behaviour.

Is that now fine to have?

Also, should this also be done per-container? Multiple containers (probably) have different `/var/vcap/sys/log/` directories. (When we were talking offline, I had thought you merged it all, instead of only grabbing it from one container)

File: container-host-files/opt/scf/bin/klog.sh [hunk 4]
```diff
@@ -5,192 +5,82 @@ set -e
...
-lookfor <<EOF
-[Warning] WSREP: no nodes coming from prim view, prim not possible
-
-Nodes all shut down improperly, MySQL must be manually rebootstrapped. Will fail
-if the IST patch has not been applied.
-EOF
+			# Get the pod logs - previous may not be there if it was successful on the first run.
+			# Unfortunately we can't get anything past the previous one
+			kubectl logs "${POD}" -n "${NS}" -c ${CONTAINER} >"${CON_DIR}/kube.log"
+			kubectl logs "${POD}" -n "${NS}" -c ${CONTAINER} >"${CON_DIR}/kube-previous.log" 2>/dev/null || true
```
Comment (mook-as) [inline container-host-files/opt/scf/bin/klog.sh]: This appears to be missing a `--previous`

File: container-host-files/opt/scf/bin/klog.sh [hunk 5]
```diff
@@ -5,192 +5,82 @@ set -e
...
 
-# mysql pod /var/vcap/sys/log/mysql/mysql.err.log
-lookfor <<EOF
-[Warning] WSREP: Failed to prepare for incremental state transfer
+	done
 
-IST addresses are misconfigured, likely a <= 4.0.1 cluster, manual
-patch must be applied and MySQL must be recovered manually.
+	# Unzip any logrotated files so the lookup can read them
+	gunzip -r "${KLOG}"
```
Comment (mook-as) [inline container-host-files/opt/scf/bin/klog.sh]: There's no `lookup` anymore?

File: container-host-files/opt/scf/bin/klog.sh (hunk 2 above)
Comment (f0rmiga) [inline container-host-files/opt/scf/bin/klog.sh]: Unnecessary change here and below. Also, you may want to set up your IDE for spaces rather than tabs.

Comment (bikramnehra) [reply container-host-files/opt/scf/bin/klog.sh]: I guess indentation is not standardized in the file, therefore even spaces won't work. For now, will revert indentation related changes and handle them in a separate commit.

Comment (bikramnehra) [reply container-host-files/opt/scf/bin/klog.sh]: For now, will revert indentation related changes and handle them in a separate commit.

File: container-host-files/opt/scf/bin/klog.sh (hunk 4 above)
Comment (bikramnehra) [reply container-host-files/opt/scf/bin/klog.sh]: Added it back, thanks for catching this.

File: container-host-files/opt/scf/bin/klog.sh (hunk 5 above)
Comment (bikramnehra) [reply container-host-files/opt/scf/bin/klog.sh]: Since we were not using it anymore, therefore, removed it. But I guess it would be a good idea to put it back for future use cases.

File: container-host-files/opt/scf/bin/klog.sh (hunk 1 above)
Comment (bikramnehra) [reply container-host-files/opt/scf/bin/klog.sh]: Sounds Legit. Reverted short options.

File: container-host-files/opt/scf/bin/klog.sh (hunk 3 above)
Comment (bikramnehra) [reply container-host-files/opt/scf/bin/klog.sh]: For #1 as discussed, we will keep it as such.

For #2, I have made changes to include `/var/vcap/sys/log/` logs from containers.

File: container-host-files/opt/scf/bin/klog.sh [hunk 6]
```diff
@@ -5,192 +5,94 @@ set -e
...
 function check_for_log_dir() {
-  kubectl exec "${POD}" --namespace "${NS}" -- bash -c "[ -d /var/vcap/sys/log ]" 2> /dev/null
+  kubectl exec "${POD}" --namespace "${NS}" --container "${CONTAINER}" -- bash -c "[ -d /var/vcap/sys/log/"${CONTAINER}" ]" 2> /dev/null
 }
 
 if [ ! -f "${DONE}" ]; then
-    rm -rf "${KLOG:?}/${NS:?}"
-    mkdir -p "${KLOG}/${NS}"
+  rm -rf "${KLOG:?}/${NS:?}"
+  BASE_DIR="${KLOG}/${NS}"
```
Comment (f0rmiga) [inline container-host-files/opt/scf/bin/klog.sh]: Please, avoid the use of the `BASE` word whenever possible as it can have different meanings. How about `NAMESPACE_DIR`?

File: container-host-files/opt/scf/bin/klog.sh [hunk 7]
```diff
@@ -5,192 +5,94 @@ set -e
...
-        fi
+    # Iterate over containers and dump logs
+    for CONTAINER in "${CONTAINERS[@]}"; do
 
-        # Get the pod logs - previous may not be there if it was successful on the first run.
-        # Unfortunately we can't get anything past the previous one
-        kubectl logs "${POD}" --namespace "${NS}" > "${DIR}/kube.log"
-        kubectl logs "${POD}" --namespace "${NS}" --previous > "${DIR}/kube-previous.log" 2> /dev/null || true
-        kubectl describe pods "${POD}" --namespace "${NS}" > "${DIR}/describe-pod.txt"
+      CON_DIR="${POD_DIR}/${CONTAINER}"
```
Comment (f0rmiga) [inline container-host-files/opt/scf/bin/klog.sh]: Could you rename `CON_DIR` to `CONTAINER_DIR`, please?

File: container-host-files/opt/scf/bin/klog.sh [hunk 8]
```diff
@@ -5,192 +5,94 @@ set -e
...
+      mkdir -p ${CON_DIR}
+
+      # Get the CF logs inside the pod if there are any
```
Comment (f0rmiga) [inline container-host-files/opt/scf/bin/klog.sh]: [nit] Since you are touching this file, could you add missing punctuation to the comments? (here and elsewhere)

File: container-host-files/opt/scf/bin/klog.sh (hunk 7 above)
Comment (bikramnehra) [reply container-host-files/opt/scf/bin/klog.sh]: DONE!

Comment (f0rmiga) [reply container-host-files/opt/scf/bin/klog.sh]: I don't see it...

File: container-host-files/opt/scf/bin/klog.sh (hunk 6 above)
Comment (bikramnehra) [reply container-host-files/opt/scf/bin/klog.sh]: DONE!

File: container-host-files/opt/scf/bin/klog.sh (hunk 7 above)
Comment (bikramnehra) [reply container-host-files/opt/scf/bin/klog.sh]: Now the changes are in the PR, you can check them

File: container-host-files/opt/scf/bin/klog.sh (hunk 8 above)
Comment (bikramnehra) [reply container-host-files/opt/scf/bin/klog.sh]: Check and see anything else to add would be useful.

File: container-host-files/opt/scf/bin/klog.sh [hunk 9]
```diff
@@ -5,192 +5,94 @@ set -e
...
+  rm -f "${DONE}" 2> /dev/null
 fi
 
 function get_phase() {
   kubectl get pod "${POD}" --namespace "${NS}" --output=jsonpath='{.status.phase}'
 }
 
 function check_for_log_dir() {
-  kubectl exec "${POD}" --namespace "${NS}" -- bash -c "[ -d /var/vcap/sys/log ]" 2> /dev/null
+  kubectl exec "${POD}" --namespace "${NS}" --container "${CONTAINER}" -- bash -c "[ -d /var/vcap/sys/log/"${CONTAINER}" ]" 2> /dev/null
```
[the remark this answers is not in the document]
Comment (bikramnehra) [reply container-host-files/opt/scf/bin/klog.sh]: DONE!

File: container-host-files/opt/scf/bin/klog.sh [hunk 10]
```diff
@@ -5,192 +5,94 @@ set -e
...
-    kubectl get events --export=true --namespace "${NS}" -o yaml > "${KLOG}/${NS}/events.yaml"
+  kubectl get all --export=true --namespace "${NS}" --output=yaml > "${KLOG}/${NS}/resources.yaml"
+  kubectl get events --export=true --namespace "${NS}" --output=yaml > "${KLOG}/${NS}/events.yaml"
 
-    touch "${DONE}"
+  touch "${DONE}"
 fi
 
 NEWLINE=0
 function lookfor {
```
[the remark this answers is not in the document]
Comment (bikramnehra) [reply container-host-files/opt/scf/bin/klog.sh]: Got rid of this method

File: container-host-files/opt/scf/bin/klog.sh [hunk 11]
```diff
@@ -5,192 +5,73 @@ set -e
...
-EOF
-
-# consul pod /var/vcap/sys/log/consul_agent/consul_agent.stdout.log
-lookfor <<EOF
-Only one node should be in bootstrap mode, not adding Raft peer.
+      # Get the pod logs - previous may not be there if it was successful on the first run.
+      # Unfortunately we can't get anything past the previous one.
+      kubectl logs "${POD}" --namespace "${NS}" --container "${CONTAINER}" > "${CONTAINER_DIR}/kube.log"
+      kubectl logs "${POD}" --namespace "${NS}" --container "${CONTAINER}" --previous > "${CONTAINER_DIR}/kube-previous.log" 2> /dev/null || true
+      kubectl describe pods "${POD}" --namespace "${NS}" > "${CONTAINER_DIR}/describe-pod.txt"
```
Comment (mook-as) [inline container-host-files/opt/scf/bin/klog.sh]: This is describing the pod (I don't think describing the container is even a thing), but for some reason we're doing this once per container.

Sorry I missed this in the last round of reviews.

Comment (bikramnehra) [reply container-host-files/opt/scf/bin/klog.sh]: Sure this seems redundant, I have moved the statement outside of container loop.

**Comment by f0rmiga** [comment]
LGTM!
